src/Data/ArrayBuffer/Typed.purs

```diff
@@ -1,87 +1,544 @@
+foreign import findImpl :: forall a b. EffectFn2 (ArrayView a) (EffectFn2 b Offset Boolean) (Nullable b)
+foreign import findIndexImpl :: forall a b. EffectFn2 (ArrayView a) (EffectFn2 b Offset Boolean) (Nullable Offset)
+foreign import indexOfImpl :: forall a b. Fn3 (ArrayView a) b (Nullable Offset) (Nullable Offset)
+foreign import lastIndexOfImpl :: forall a b. Fn3 (ArrayView a) b (Nullable Offset) (Nullable Offset)
+
+
+-- | Value-oriented array offset
+type Offset = Int
+-- | Value-oriented array length
+type Length = Int
+
+-- | Represents a range of indicies, where if omitted, it represents the whole span.
+-- | If only the second argument is omitted, then it represents the remainder of the span after the first index.
+type Range = Maybe (Tuple Offset (Maybe Offset))
+
+
+-- TODO use purescript-quotient
+-- | Typeclass that associates a measured user-level type with a typed array.
+-- |
+-- | #### Creation
+-- |
+-- | - `whole`, `remainder`, and `part` are methods for building a typed array accessible interface
+-- |   on top of an existing `ArrayBuffer` - Note, `part` and `remainder` may behave unintuitively -
+-- |   when the operation is isomorphic to `whole`, the new TypedArray uses the same buffer as the input,
+-- |   but not when the portion is a sub-array of the original buffer, a new one is made with
+-- |   `Data.ArrayBuffer.ArrayBuffer.slice`.
+-- | - `empty` and `fromArray` are methods for creating pure typed arrays
+-- |
+-- | #### Modification
+-- |
+-- | - `fill`, `set`, and `setTyped` are methods for assigning values from external sources
+-- | - `map` and `traverse` allow you to create a new array from the existing values in another
+-- | - `copyWithin` allows you to set values to the array that exist in other parts of the array
+-- | - `filter` creates a new array without the values that don't pass a predicate
+-- | - `reverse` modifies an existing array in-place, with all values reversed
+-- | - `sort` modifies an existing array in-place, with all values sorted
+-- |
+-- | #### Access
+-- |
+-- | - `elem`, `all`, and `any` are functions for testing the contents of an array
+-- | - `unsafeAt`, `hasIndex`, and `at` are used to get values from an array, with an offset
+-- | - `foldr`, `foldrM`, `foldr1`, `foldr1M`, `foldl`, `foldlM`, `foldl1`, `foldl1M` all can reduce an array
+-- | - `find` and `findIndex` are searching functions via a predicate
+-- | - `indexOf` and `lastIndexOf` are searching functions via equality
+-- | - `slice` returns a new typed array on the same array buffer content as the input
+-- | - `subArray` returns a new typed array with a separate array buffer
+-- | - `toString` prints to a CSV, `toString'` allows you to supply the delimiter
+-- | - `toArray` returns an array of numeric values
+class BinaryValue a t <= TypedArray (a :: ArrayViewType) (t :: Type) | a -> t where
+  -- | View mapping the whole `ArrayBuffer`.
+  whole :: ArrayBuffer -> ArrayView a
```

Comment:
I guess we should have `whole`/`remainder`/`part` share the same return type. Did you keep it this way to be backwards-compatible?

Reply:
I left `whole` to be pure, because it can't fail, but `remainder` and `part` can raise exceptions because of range errors. It's symmetric across the different modules, though.